geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LPopDUnitTest.java

```diff
@@ -0,0 +1,195 @@
+
+import static org.apache.geode.test.dunit.rules.RedisClusterStartupRule.BIND_ADDRESS;
+import static org.apache.geode.test.dunit.rules.RedisClusterStartupRule.REDIS_CLIENT_TIMEOUT;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+
+import org.apache.geode.test.awaitility.GeodeAwaitility;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+import org.apache.geode.test.junit.rules.ExecutorServiceRule;
+
+public class LPopDUnitTest {
+  public static final int MINIMUM_ITERATIONS = 10000;
+  private static MemberVM locator;
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule();
+
+  @Rule
+  public ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  private static JedisCluster jedis;
+
+  @BeforeClass
+  public static void classSetup() {
+    locator = clusterStartUp.startLocatorVM(0);
+    clusterStartUp.startRedisVM(2, locator.getPort());
+    clusterStartUp.startRedisVM(3, locator.getPort());
+  }
+
+  @Before
+  public void testSetup() {
+    clusterStartUp.startRedisVM(1, locator.getPort());
+    int redisServerPort = clusterStartUp.getRedisPort(1);
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort), REDIS_CLIENT_TIMEOUT);
+    clusterStartUp.flushAll();
+  }
+
+  @After
+  public void tearDown() {
+    jedis.close();
+  }
+
+  @Test
+  public void shouldDistributeDataAmongCluster_andRetainDataAfterServerCrash() {
+    String key = makeListKeyWithHashtag(1, clusterStartUp.getKeyOnServer("lpop", 1));
+    List<String> elementList = makeElementList(key, MINIMUM_ITERATIONS);
+    lpushPerformAndVerify(key, elementList);
+
+    // Remove all but last element
+    for (int i = MINIMUM_ITERATIONS - 1; i > 0; i--) {
+      assertThat(jedis.lpop(key)).isEqualTo(makeElementString(key, i));
+    }
+    clusterStartUp.crashVM(1); // kill primary server
+
+    assertThat(jedis.lpop(key)).isEqualTo(makeElementString(key, 0));
+    assertThat(jedis.exists(key)).isFalse();
+  }
+
+  @Test
+  public void givenBucketsMoveDuringLpop_thenOperationsAreNotLost() throws Exception {
+    AtomicLong runningCount = new AtomicLong(3);
+    List<String> listHashtags = makeListHashtags();
+    List<String> keys = makeListKeys(listHashtags);
+
+    List<String> elementList1 = makeElementList(keys.get(0), MINIMUM_ITERATIONS);
+    List<String> elementList2 = makeElementList(keys.get(1), MINIMUM_ITERATIONS);
+    List<String> elementList3 = makeElementList(keys.get(2), MINIMUM_ITERATIONS);
```

Comment:
The name `MINIMUM_ITERATIONS` is a bit misleading, since we actually add exactly that many elements to the list every time. It should probably be renamed to something like "INITIAL_LIST_SIZE" and any assertions on it using `.isGreaterThanOrEqualTo()` should be changed to `.isEqualTo()`

Reply:
Renamed. .isGreaterThanOrEqualTo() were from earlier iteration shuffling buckets when building lists, but that's not this test's concern.